frontend/src/sagas/auth.js

```diff
@@ -0,0 +1,152 @@
+    try {
+        let response;
+        if (payload.type === GOOGLE) {
+            response = { data: { token: payload.token, email: '' } };
+        } else {
+            response = yield call(axiosCall, LOGIN_URL, 'POST', payload.result);
+        }
+        const { token, email } = response.data;
+        const decodedJWT = jwtDecode(token);
+        const expirationDate = new Date(decodedJWT.exp * 1000);
+
+        axios.defaults.headers.common.Authorization = TOKEN_BEGIN + token;
+
+        yield localStorage.setItem('token', TOKEN_BEGIN + token);
+        yield localStorage.setItem('expirationDate', expirationDate);
+        yield localStorage.setItem('userRole', decodedJWT.roles);
+        yield localStorage.setItem('email', email || decodedJWT.sub);
+
+        yield put(authSuccess({ token, role: decodedJWT.roles, email }));
+        yield put(authAutoLogout(decodedJWT.exp * 1000 - new Date().getTime()));
+    } catch (error) {
+        yield put(
+            setAuthError({
+                login: error.response ? error.response.data.message : i18n.t(COMMON_ERROR_MESSAGE),
+            }),
+        );
+    } finally {
+        yield put(setLoading(false));
+    }
+}
+
+function* registerAccount(payload) {
+    try {
+        const response = yield call(axiosCall, REGISTRATION_URL, 'POST', payload.result);
+        yield put(registerUserSuccess(response));
+    } catch (error) {
+        yield put(
+            setAuthError({
+                registration: error.response
+                    ? error.response.data.message
+                    : i18n.t(COMMON_ERROR_MESSAGE),
+            }),
+        );
+    } finally {
+        yield put(setLoading(false));
+    }
+}
+
+function* logoutOfAccount(payload) {
+    try {
+        yield call(axiosCall, LOGOUT_URL, 'POST', payload);
+        yield put(logout());
+    } catch (error) {
+        yield put(
+            setAuthError({
+                login: error.response ? error.response.data.message : i18n.t(COMMON_ERROR_MESSAGE),
+            }),
+        );
+    }
+}
+
+function* checkAuthState(payload) {
+    const { token, role } = payload;
+    const expirationDate = new Date(localStorage.getItem('expirationDate'));
+    if (!token || expirationDate <= new Date()) {
+        yield put(logout());
+    } else {
+        yield put(authSuccess({ token, role }));
+    }
+}
+
+function* checkAuthTimeout(action) {
+    yield delay(action.expirationTime);
+    yield call(logoutOfAccount);
+}
+
+function* activateUserAccount(payload) {
+    try {
+        const response = yield call(
+            axiosCall,
+            `${ACTIVATE_ACCOUNT_URL}?token=${payload.result}`,
+            'PUT',
+        );
+        yield put(activateSuccess(response));
+    } catch (error) {
+        yield put(
+            setAuthError({
+                activationError: error.response
+                    ? error.response.data.message
+                    : i18n.t(COMMON_ERROR_MESSAGE),
```

Comment:
can you do something with repeated messages?

Reply:
now I use createErrorMessage method